refactor!: Virtual DocType

ankush opened the pull request:
- [x] fix form view
- [x] parent comments via https://github.com/frappe/frappe/pull/17588
- [x] document all the expectations from virtual doctypes in form of protocol class. 
- [x] hide test virtual doctype in prod
- [x] remove "Test rename new" (some random testing doctype committed in code)
- [x] Fix apis like get_list (likely a breaking change in virtual doctypes)
- [x] add a realistic example in tests. 
- [x] support delete_doc (from desk and API)



**Breaking Change**

- get_list, get_count, get_stats need to be static methods on virtual doctype controller. 
- db_insert and db_update need to accept arbitrary args/kwargs to let framework extend these methods when required. 


New interface requirements for virtual doctype documented in https://github.com/frappe/frappe/blob/develop/frappe/model/virtual_doctype.py file. 


deferred to push other imp change before v14:
- [ ] Validate and warn if protocol expectations aren't met!
- [ ] clearly document what parts of ORM work with virtual doctypes
- [ ] Desk UX (lots of actions are broken / should be hidden for virtual doctypes)

planning to resolve: https://github.com/frappe/frappe/pull/17188, https://github.com/frappe/frappe/issues/13302, https://github.com/frappe/frappe/issues/14950

ankush committed:
fix: form view broken for virtual doctype

etobella commented:
Once you create a virtual doctype if you press on Filter By Assigned To the following error is raised
### App Versions
```
{
	"demo": "0.0.1",
	"frappe": "14.0.0-dev"
}
```
### Route
```
List/demo/List
```
### Trackeback
```
Traceback (most recent call last):
  File "apps/frappe/frappe/app.py", line 68, in application
    response = frappe.api.handle()
  File "apps/frappe/frappe/api.py", line 54, in handle
    return frappe.handler.handle()
  File "apps/frappe/frappe/handler.py", line 45, in handle
    data = execute_cmd(cmd)
  File "apps/frappe/frappe/handler.py", line 83, in execute_cmd
    return frappe.call(method, **frappe.form_dict)
  File "apps/frappe/frappe/__init__.py", line 1579, in call
    return fn(*args, **newargs)
  File "apps/frappe/frappe/desk/listview.py", line 42, in get_group_by_count
    frappe.qb.from_(ToDo)
  File "apps/frappe/frappe/query_builder/utils.py", line 76, in execute_query
    return frappe.db.sql(query, params, *args, **kwargs)  # nosemgrep
  File "apps/frappe/frappe/database/database.py", line 175, in sql
    self._cursor.execute(query, values)
  File "env/lib/python3.10/site-packages/pymysql/cursors.py", line 148, in execute
    result = self._query(query)
  File "env/lib/python3.10/site-packages/pymysql/cursors.py", line 310, in _query
    conn.query(q)
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 548, in query
    self._affected_rows = self._read_query_result(unbuffered=unbuffered)
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 775, in _read_query_result
    result.read()
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 1156, in read
    first_packet = self.connection._read_packet()
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 725, in _read_packet
    packet.raise_for_error()
  File "env/lib/python3.10/site-packages/pymysql/protocol.py", line 221, in raise_for_error
    err.raise_mysql_exception(self._data)
  File "env/lib/python3.10/site-packages/pymysql/err.py", line 143, in raise_mysql_exception
    raise errorclass(errno, errval)
pymysql.err.ProgrammingError: (1146, "Table '_9f1c4df770ef4359.tabdemo' doesn't exist")

```
### Request Data
```
{
	"type": "POST",
	"args": {
		"doctype": "demo",
		"current_filters": "[]",
		"field": "assigned_to"
	},
	"headers": {},
	"error_handlers": {},
	"url": "/api/method/frappe.desk.listview.get_group_by_count"
}
```
### Response Data
```
{
	"exception": "pymysql.err.ProgrammingError: (1146, \"Table '_9f1c4df770ef4359.tabdemo' doesn't exist\")"
}
```


Same happens if you press on the filter Created by

### App Versions
```
{
	"demo": "0.0.1",
	"frappe": "14.0.0-dev"
}
```
### Route
```
List/demo/List
```
### Trackeback
```
Traceback (most recent call last):
  File "apps/frappe/frappe/app.py", line 68, in application
    response = frappe.api.handle()
  File "apps/frappe/frappe/api.py", line 54, in handle
    return frappe.handler.handle()
  File "apps/frappe/frappe/handler.py", line 45, in handle
    data = execute_cmd(cmd)
  File "apps/frappe/frappe/handler.py", line 83, in execute_cmd
    return frappe.call(method, **frappe.form_dict)
  File "apps/frappe/frappe/__init__.py", line 1579, in call
    return fn(*args, **newargs)
  File "apps/frappe/frappe/desk/listview.py", line 42, in get_group_by_count
    frappe.qb.from_(ToDo)
  File "apps/frappe/frappe/query_builder/utils.py", line 76, in execute_query
    return frappe.db.sql(query, params, *args, **kwargs)  # nosemgrep
  File "apps/frappe/frappe/database/database.py", line 175, in sql
    self._cursor.execute(query, values)
  File "env/lib/python3.10/site-packages/pymysql/cursors.py", line 148, in execute
    result = self._query(query)
  File "env/lib/python3.10/site-packages/pymysql/cursors.py", line 310, in _query
    conn.query(q)
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 548, in query
    self._affected_rows = self._read_query_result(unbuffered=unbuffered)
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 775, in _read_query_result
    result.read()
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 1156, in read
    first_packet = self.connection._read_packet()
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 725, in _read_packet
    packet.raise_for_error()
  File "env/lib/python3.10/site-packages/pymysql/protocol.py", line 221, in raise_for_error
    err.raise_mysql_exception(self._data)
  File "env/lib/python3.10/site-packages/pymysql/err.py", line 143, in raise_mysql_exception
    raise errorclass(errno, errval)
pymysql.err.ProgrammingError: (1146, "Table '_9f1c4df770ef4359.tabdemo' doesn't exist")

```
### Request Data
```
{
	"type": "POST",
	"args": {
		"doctype": "demo",
		"current_filters": "[]",
		"field": "assigned_to"
	},
	"headers": {},
	"error_handlers": {},
	"url": "/api/method/frappe.desk.listview.get_group_by_count"
}
```
### Response Data
```
{
	"exception": "pymysql.err.ProgrammingError: (1146, \"Table '_9f1c4df770ef4359.tabdemo' doesn't exist\")"
}
```

For this reason, this code existed:  https://github.com/frappe/frappe/pull/17188/files#diff-0c820faa40e5ac07693b83f29d8505e63b5e46c8fa6019a5239da3b89144f531L1832


If you press on the heart of the document also an error is raised:

### App Versions
```
{
	"demo": "0.0.1",
	"frappe": "14.0.0-dev"
}
```
### Route
```
List/demo/List
```
### Trackeback
```
Traceback (most recent call last):
  File "apps/frappe/frappe/app.py", line 68, in application
    response = frappe.api.handle()
  File "apps/frappe/frappe/api.py", line 54, in handle
    return frappe.handler.handle()
  File "apps/frappe/frappe/handler.py", line 45, in handle
    data = execute_cmd(cmd)
  File "apps/frappe/frappe/handler.py", line 83, in execute_cmd
    return frappe.call(method, **frappe.form_dict)
  File "apps/frappe/frappe/__init__.py", line 1579, in call
    return fn(*args, **newargs)
  File "apps/frappe/frappe/desk/listview.py", line 42, in get_group_by_count
    frappe.qb.from_(ToDo)
  File "apps/frappe/frappe/query_builder/utils.py", line 76, in execute_query
    return frappe.db.sql(query, params, *args, **kwargs)  # nosemgrep
  File "apps/frappe/frappe/database/database.py", line 175, in sql
    self._cursor.execute(query, values)
  File "env/lib/python3.10/site-packages/pymysql/cursors.py", line 148, in execute
    result = self._query(query)
  File "env/lib/python3.10/site-packages/pymysql/cursors.py", line 310, in _query
    conn.query(q)
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 548, in query
    self._affected_rows = self._read_query_result(unbuffered=unbuffered)
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 775, in _read_query_result
    result.read()
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 1156, in read
    first_packet = self.connection._read_packet()
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 725, in _read_packet
    packet.raise_for_error()
  File "env/lib/python3.10/site-packages/pymysql/protocol.py", line 221, in raise_for_error
    err.raise_mysql_exception(self._data)
  File "env/lib/python3.10/site-packages/pymysql/err.py", line 143, in raise_mysql_exception
    raise errorclass(errno, errval)
pymysql.err.ProgrammingError: (1146, "Table '_9f1c4df770ef4359.tabdemo' doesn't exist")

```
### Request Data
```
{
	"type": "POST",
	"args": {
		"doctype": "demo",
		"current_filters": "[]",
		"field": "assigned_to"
	},
	"headers": {},
	"error_handlers": {},
	"url": "/api/method/frappe.desk.listview.get_group_by_count"
}
```
### Response Data
```
{
	"exception": "pymysql.err.ProgrammingError: (1146, \"Table '_9f1c4df770ef4359.tabdemo' doesn't exist\")"
}
```

If you try to add a tag another error is raised
### App Versions
```
{
	"demo": "0.0.1",
	"frappe": "14.0.0-dev"
}
```
### Route
```
Form/demo/4b05caab6e
```
### Trackeback
```
Traceback (most recent call last):
  File "apps/frappe/frappe/app.py", line 68, in application
    response = frappe.api.handle()
  File "apps/frappe/frappe/api.py", line 54, in handle
    return frappe.handler.handle()
  File "apps/frappe/frappe/handler.py", line 45, in handle
    data = execute_cmd(cmd)
  File "apps/frappe/frappe/handler.py", line 83, in execute_cmd
    return frappe.call(method, **frappe.form_dict)
  File "apps/frappe/frappe/__init__.py", line 1579, in call
    return fn(*args, **newargs)
  File "apps/frappe/frappe/desk/listview.py", line 42, in get_group_by_count
    frappe.qb.from_(ToDo)
  File "apps/frappe/frappe/query_builder/utils.py", line 76, in execute_query
    return frappe.db.sql(query, params, *args, **kwargs)  # nosemgrep
  File "apps/frappe/frappe/database/database.py", line 175, in sql
    self._cursor.execute(query, values)
  File "env/lib/python3.10/site-packages/pymysql/cursors.py", line 148, in execute
    result = self._query(query)
  File "env/lib/python3.10/site-packages/pymysql/cursors.py", line 310, in _query
    conn.query(q)
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 548, in query
    self._affected_rows = self._read_query_result(unbuffered=unbuffered)
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 775, in _read_query_result
    result.read()
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 1156, in read
    first_packet = self.connection._read_packet()
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 725, in _read_packet
    packet.raise_for_error()
  File "env/lib/python3.10/site-packages/pymysql/protocol.py", line 221, in raise_for_error
    err.raise_mysql_exception(self._data)
  File "env/lib/python3.10/site-packages/pymysql/err.py", line 143, in raise_mysql_exception
    raise errorclass(errno, errval)
pymysql.err.ProgrammingError: (1146, "Table '_9f1c4df770ef4359.tabdemo' doesn't exist")

```
### Request Data
```
{
	"type": "POST",
	"args": {
		"doctype": "demo",
		"current_filters": "[]",
		"field": "assigned_to"
	},
	"headers": {},
	"error_handlers": {},
	"url": "/api/method/frappe.desk.listview.get_group_by_count"
}
```
### Response Data
```
{
	"exception": "pymysql.err.ProgrammingError: (1146, \"Table '_9f1c4df770ef4359.tabdemo' doesn't exist\")"
}
```


If you try to assign the document to someone also it fails:


### App Versions
```
{
	"demo": "0.0.1",
	"frappe": "14.0.0-dev"
}
```
### Route
```
Form/demo/4b05caab6e
```
### Trackeback
```
Traceback (most recent call last):
  File "apps/frappe/frappe/app.py", line 68, in application
    response = frappe.api.handle()
  File "apps/frappe/frappe/api.py", line 54, in handle
    return frappe.handler.handle()
  File "apps/frappe/frappe/handler.py", line 45, in handle
    data = execute_cmd(cmd)
  File "apps/frappe/frappe/handler.py", line 83, in execute_cmd
    return frappe.call(method, **frappe.form_dict)
  File "apps/frappe/frappe/__init__.py", line 1579, in call
    return fn(*args, **newargs)
  File "apps/frappe/frappe/desk/listview.py", line 42, in get_group_by_count
    frappe.qb.from_(ToDo)
  File "apps/frappe/frappe/query_builder/utils.py", line 76, in execute_query
    return frappe.db.sql(query, params, *args, **kwargs)  # nosemgrep
  File "apps/frappe/frappe/database/database.py", line 175, in sql
    self._cursor.execute(query, values)
  File "env/lib/python3.10/site-packages/pymysql/cursors.py", line 148, in execute
    result = self._query(query)
  File "env/lib/python3.10/site-packages/pymysql/cursors.py", line 310, in _query
    conn.query(q)
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 548, in query
    self._affected_rows = self._read_query_result(unbuffered=unbuffered)
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 775, in _read_query_result
    result.read()
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 1156, in read
    first_packet = self.connection._read_packet()
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 725, in _read_packet
    packet.raise_for_error()
  File "env/lib/python3.10/site-packages/pymysql/protocol.py", line 221, in raise_for_error
    err.raise_mysql_exception(self._data)
  File "env/lib/python3.10/site-packages/pymysql/err.py", line 143, in raise_mysql_exception
    raise errorclass(errno, errval)
pymysql.err.ProgrammingError: (1146, "Table '_9f1c4df770ef4359.tabdemo' doesn't exist")

```
### Request Data
```
{
	"type": "POST",
	"args": {
		"doctype": "demo",
		"current_filters": "[]",
		"field": "assigned_to"
	},
	"headers": {},
	"error_handlers": {},
	"url": "/api/method/frappe.desk.listview.get_group_by_count"
}
```
### Response Data
```
{
	"exception": "pymysql.err.ProgrammingError: (1146, \"Table '_9f1c4df770ef4359.tabdemo' doesn't exist\")"
}
```

etobella commented:
Same happens if you press on the filter Created by

### App Versions
```
{
	"demo": "0.0.1",
	"frappe": "14.0.0-dev"
}
```
### Route
```
List/demo/List
```
### Trackeback
```
Traceback (most recent call last):
  File "apps/frappe/frappe/app.py", line 68, in application
    response = frappe.api.handle()
  File "apps/frappe/frappe/api.py", line 54, in handle
    return frappe.handler.handle()
  File "apps/frappe/frappe/handler.py", line 45, in handle
    data = execute_cmd(cmd)
  File "apps/frappe/frappe/handler.py", line 83, in execute_cmd
    return frappe.call(method, **frappe.form_dict)
  File "apps/frappe/frappe/__init__.py", line 1579, in call
    return fn(*args, **newargs)
  File "apps/frappe/frappe/desk/listview.py", line 42, in get_group_by_count
    frappe.qb.from_(ToDo)
  File "apps/frappe/frappe/query_builder/utils.py", line 76, in execute_query
    return frappe.db.sql(query, params, *args, **kwargs)  # nosemgrep
  File "apps/frappe/frappe/database/database.py", line 175, in sql
    self._cursor.execute(query, values)
  File "env/lib/python3.10/site-packages/pymysql/cursors.py", line 148, in execute
    result = self._query(query)
  File "env/lib/python3.10/site-packages/pymysql/cursors.py", line 310, in _query
    conn.query(q)
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 548, in query
    self._affected_rows = self._read_query_result(unbuffered=unbuffered)
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 775, in _read_query_result
    result.read()
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 1156, in read
    first_packet = self.connection._read_packet()
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 725, in _read_packet
    packet.raise_for_error()
  File "env/lib/python3.10/site-packages/pymysql/protocol.py", line 221, in raise_for_error
    err.raise_mysql_exception(self._data)
  File "env/lib/python3.10/site-packages/pymysql/err.py", line 143, in raise_mysql_exception
    raise errorclass(errno, errval)
pymysql.err.ProgrammingError: (1146, "Table '_9f1c4df770ef4359.tabdemo' doesn't exist")

```
### Request Data
```
{
	"type": "POST",
	"args": {
		"doctype": "demo",
		"current_filters": "[]",
		"field": "assigned_to"
	},
	"headers": {},
	"error_handlers": {},
	"url": "/api/method/frappe.desk.listview.get_group_by_count"
}
```
### Response Data
```
{
	"exception": "pymysql.err.ProgrammingError: (1146, \"Table '_9f1c4df770ef4359.tabdemo' doesn't exist\")"
}
```

For this reason, this code existed:  https://github.com/frappe/frappe/pull/17188/files#diff-0c820faa40e5ac07693b83f29d8505e63b5e46c8fa6019a5239da3b89144f531L1832

etobella commented:
If you press on the heart of the document also an error is raised:

### App Versions
```
{
	"demo": "0.0.1",
	"frappe": "14.0.0-dev"
}
```
### Route
```
List/demo/List
```
### Trackeback
```
Traceback (most recent call last):
  File "apps/frappe/frappe/app.py", line 68, in application
    response = frappe.api.handle()
  File "apps/frappe/frappe/api.py", line 54, in handle
    return frappe.handler.handle()
  File "apps/frappe/frappe/handler.py", line 45, in handle
    data = execute_cmd(cmd)
  File "apps/frappe/frappe/handler.py", line 83, in execute_cmd
    return frappe.call(method, **frappe.form_dict)
  File "apps/frappe/frappe/__init__.py", line 1579, in call
    return fn(*args, **newargs)
  File "apps/frappe/frappe/desk/listview.py", line 42, in get_group_by_count
    frappe.qb.from_(ToDo)
  File "apps/frappe/frappe/query_builder/utils.py", line 76, in execute_query
    return frappe.db.sql(query, params, *args, **kwargs)  # nosemgrep
  File "apps/frappe/frappe/database/database.py", line 175, in sql
    self._cursor.execute(query, values)
  File "env/lib/python3.10/site-packages/pymysql/cursors.py", line 148, in execute
    result = self._query(query)
  File "env/lib/python3.10/site-packages/pymysql/cursors.py", line 310, in _query
    conn.query(q)
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 548, in query
    self._affected_rows = self._read_query_result(unbuffered=unbuffered)
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 775, in _read_query_result
    result.read()
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 1156, in read
    first_packet = self.connection._read_packet()
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 725, in _read_packet
    packet.raise_for_error()
  File "env/lib/python3.10/site-packages/pymysql/protocol.py", line 221, in raise_for_error
    err.raise_mysql_exception(self._data)
  File "env/lib/python3.10/site-packages/pymysql/err.py", line 143, in raise_mysql_exception
    raise errorclass(errno, errval)
pymysql.err.ProgrammingError: (1146, "Table '_9f1c4df770ef4359.tabdemo' doesn't exist")

```
### Request Data
```
{
	"type": "POST",
	"args": {
		"doctype": "demo",
		"current_filters": "[]",
		"field": "assigned_to"
	},
	"headers": {},
	"error_handlers": {},
	"url": "/api/method/frappe.desk.listview.get_group_by_count"
}
```
### Response Data
```
{
	"exception": "pymysql.err.ProgrammingError: (1146, \"Table '_9f1c4df770ef4359.tabdemo' doesn't exist\")"
}
```

etobella commented:
If you try to add a tag another error is raised
### App Versions
```
{
	"demo": "0.0.1",
	"frappe": "14.0.0-dev"
}
```
### Route
```
Form/demo/4b05caab6e
```
### Trackeback
```
Traceback (most recent call last):
  File "apps/frappe/frappe/app.py", line 68, in application
    response = frappe.api.handle()
  File "apps/frappe/frappe/api.py", line 54, in handle
    return frappe.handler.handle()
  File "apps/frappe/frappe/handler.py", line 45, in handle
    data = execute_cmd(cmd)
  File "apps/frappe/frappe/handler.py", line 83, in execute_cmd
    return frappe.call(method, **frappe.form_dict)
  File "apps/frappe/frappe/__init__.py", line 1579, in call
    return fn(*args, **newargs)
  File "apps/frappe/frappe/desk/listview.py", line 42, in get_group_by_count
    frappe.qb.from_(ToDo)
  File "apps/frappe/frappe/query_builder/utils.py", line 76, in execute_query
    return frappe.db.sql(query, params, *args, **kwargs)  # nosemgrep
  File "apps/frappe/frappe/database/database.py", line 175, in sql
    self._cursor.execute(query, values)
  File "env/lib/python3.10/site-packages/pymysql/cursors.py", line 148, in execute
    result = self._query(query)
  File "env/lib/python3.10/site-packages/pymysql/cursors.py", line 310, in _query
    conn.query(q)
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 548, in query
    self._affected_rows = self._read_query_result(unbuffered=unbuffered)
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 775, in _read_query_result
    result.read()
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 1156, in read
    first_packet = self.connection._read_packet()
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 725, in _read_packet
    packet.raise_for_error()
  File "env/lib/python3.10/site-packages/pymysql/protocol.py", line 221, in raise_for_error
    err.raise_mysql_exception(self._data)
  File "env/lib/python3.10/site-packages/pymysql/err.py", line 143, in raise_mysql_exception
    raise errorclass(errno, errval)
pymysql.err.ProgrammingError: (1146, "Table '_9f1c4df770ef4359.tabdemo' doesn't exist")

```
### Request Data
```
{
	"type": "POST",
	"args": {
		"doctype": "demo",
		"current_filters": "[]",
		"field": "assigned_to"
	},
	"headers": {},
	"error_handlers": {},
	"url": "/api/method/frappe.desk.listview.get_group_by_count"
}
```
### Response Data
```
{
	"exception": "pymysql.err.ProgrammingError: (1146, \"Table '_9f1c4df770ef4359.tabdemo' doesn't exist\")"
}
```

etobella commented:
If you try to assign the document to someone also it fails:


### App Versions
```
{
	"demo": "0.0.1",
	"frappe": "14.0.0-dev"
}
```
### Route
```
Form/demo/4b05caab6e
```
### Trackeback
```
Traceback (most recent call last):
  File "apps/frappe/frappe/app.py", line 68, in application
    response = frappe.api.handle()
  File "apps/frappe/frappe/api.py", line 54, in handle
    return frappe.handler.handle()
  File "apps/frappe/frappe/handler.py", line 45, in handle
    data = execute_cmd(cmd)
  File "apps/frappe/frappe/handler.py", line 83, in execute_cmd
    return frappe.call(method, **frappe.form_dict)
  File "apps/frappe/frappe/__init__.py", line 1579, in call
    return fn(*args, **newargs)
  File "apps/frappe/frappe/desk/listview.py", line 42, in get_group_by_count
    frappe.qb.from_(ToDo)
  File "apps/frappe/frappe/query_builder/utils.py", line 76, in execute_query
    return frappe.db.sql(query, params, *args, **kwargs)  # nosemgrep
  File "apps/frappe/frappe/database/database.py", line 175, in sql
    self._cursor.execute(query, values)
  File "env/lib/python3.10/site-packages/pymysql/cursors.py", line 148, in execute
    result = self._query(query)
  File "env/lib/python3.10/site-packages/pymysql/cursors.py", line 310, in _query
    conn.query(q)
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 548, in query
    self._affected_rows = self._read_query_result(unbuffered=unbuffered)
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 775, in _read_query_result
    result.read()
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 1156, in read
    first_packet = self.connection._read_packet()
  File "env/lib/python3.10/site-packages/pymysql/connections.py", line 725, in _read_packet
    packet.raise_for_error()
  File "env/lib/python3.10/site-packages/pymysql/protocol.py", line 221, in raise_for_error
    err.raise_mysql_exception(self._data)
  File "env/lib/python3.10/site-packages/pymysql/err.py", line 143, in raise_mysql_exception
    raise errorclass(errno, errval)
pymysql.err.ProgrammingError: (1146, "Table '_9f1c4df770ef4359.tabdemo' doesn't exist")

```
### Request Data
```
{
	"type": "POST",
	"args": {
		"doctype": "demo",
		"current_filters": "[]",
		"field": "assigned_to"
	},
	"headers": {},
	"error_handlers": {},
	"url": "/api/method/frappe.desk.listview.get_group_by_count"
}
```
### Response Data
```
{
	"exception": "pymysql.err.ProgrammingError: (1146, \"Table '_9f1c4df770ef4359.tabdemo' doesn't exist\")"
}
```

ankush commented:
@etobella some of those errors are just not expected to work. 

IMO anything that directly inspects DB is not expected to work, maybe we need to hide some of these things from UI. The purpose of the virtual doc is to provide a form and list view foo simple CRUD-like operations on some other data store. 

I get it that it's not explicitly documented but expecting more than that with current design is just not feasible. 

More functionality can be added over time but it has to be implemented by the doctype controller itself and can't be handled by the framework magically.

etobella commented:
I know it should not be expected, but functionality was there and users might get errors if they tried

shridarpatil committed:
fix: do not fetch comments from parent for virtual doctype

ankush committed:
refactor: move is_virtual_doctype to relevant file

ankush committed:
chore: remove test doctypes from prod

ankush committed:
refactor!: better API contracts for virtual doctype

Current APIs implement class methods as instance method, which is
problamamtic while implementing methods. E.g. If load_from_db doesn't
like empty docname then all class method will stop working.

This change while breaking is essential for usability of virtual
doctype.

ankush committed:
refactor: add reload function back

Assigning a function to a different name breaks inheritance model.

E.g. doc.reload() won't call virtual doctype's load_from_db but call
original load_from_db

ankush committed:
test: add tests for virtual doctype desk interactions

ankush committed:
feat: delete support for virtual doctypes from desk

gavindsouza commented on frappe/core/doctype/test/test.py:
```diff
@@ -1,43 +1,74 @@
+
+def update_data(data: Data) -> None:
+	"""Flush updated data to disk"""
+	with open(DATA_FILE, "w+") as data_file:
```
this should be pushed into `os.path.join(frappe.local.site, DATA_FILE)` instead of the base sites folder?

gavindsouza commented on frappe/core/doctype/test/test.json:
```diff
@@ -36,7 +36,9 @@
+ "read_only": 1,
  "sort_field": "modified",
  "sort_order": "DESC",
+ "states": [],
```
Should we rename this to "Test Virtual DocType" instead? As "test", it looks like yet another directory that has been tracked as a blunder. 😆

shridarpatil replied:
Ideally, this should not be in the core doctype. I guess it was committed unknowingly.

gavindsouza commented on frappe/core/doctype/comment/comment.py:
```diff
@@ -11,6 +11,7 @@
 	get_title,
 	get_title_html,
 )
+from frappe.desk.reportview import is_virtual_doctype
```
Simplify import path? Will improve runtime sanity a wee bit IMO

```suggestion
from frappe.model.utils import is_virtual_doctype
```

gavindsouza commented on frappe/desk/form/load.py:
```diff
@@ -30,7 +31,7 @@ def getdoc(doctype, name, user=None):
 		name = doctype
 
-	if not frappe.db.exists(doctype, name):
+	if not frappe.db.exists(doctype, name) and not is_virtual_doctype(doctype):
```
`is_vitrtual_doctype` is likely site cached (and thus faster than a query)? Should we reverse the order? 

Ignore if you think it doesn't matter much

gavindsouza commented:
Broadly, LGTM. Changes may be breaking but this brings some (design) sanity which is highly appreciated.

ankush committed:
refactor: simplify virtual doctype example

ankush committed:
fix: correct import path

Co-authored-by: gavin <gavin18d@gmail.com>
